crates/holochain/src/sweettest/sweet_topos/network.rs

```diff
@@ -0,0 +1,509 @@
+    /// make sure that we don't create self edges or duplicate edges, and we
+    /// need to make sure that we don't create edges that have an origin node
+    /// referencing a different conductor.
+    #[shrinkwrap(main_field)]
+    graph: Graph<NetworkTopologyNode, NetworkTopologyEdge, Directed, usize>,
+}
+
+/// Errors that can occur when manipulating a `NetworkTopology`.
+#[derive(derive_more::Error, derive_more::Display, Debug)]
+pub enum NetworkTopologyError {
+    /// Failed to remove an edge from the graph.
+    #[display(fmt = "Failed to remove an edge from the graph.")]
+    EdgeRemove,
+    /// Failed to calculate the smallest partition.
+    #[display(fmt = "Failed to calculate the smallest partition.")]
+    UnknownSmallestPartition,
+    /// The smallest partition is empty (no nodes).
+    #[display(fmt = "The smallest partition is empty (no nodes).")]
+    EmptySmallestPartition,
+    /// The node index is not in the graph.
+    #[display(fmt = "The node index is not in the graph.")]
+    DanglingNodeIndex,
+    /// The edge is a self edge.
+    #[display(fmt = "The edge is a self edge.")]
+    IntegritySelfEdge,
+    /// The edge is a duplicate edge.
+    #[display(fmt = "The edge is a duplicate edge.")]
+    IntegrityDuplicateEdge,
+    /// The edge source is not in the graph.
+    #[display(fmt = "The edge source is not in the graph.")]
+    IntegrityDanglingEdgeSource,
+    /// The edge target is not in the graph.
+    #[display(fmt = "The edge target is not in the graph.")]
+    IntegrityDanglingEdgeTarget,
+    /// The edge source conductor does not match the edge source.
+    #[display(fmt = "The edge source conductor does not match the edge source.")]
+    IntegritySourceConductorMismatch,
+    /// The edge target conductor does not match the edge target.
+    #[display(fmt = "The edge target conductor does not match the edge target.")]
+    IntegrityTargetConductorMismatch,
+}
+
+impl From<NetworkTopologyError> for MutationError {
+    fn from(e: NetworkTopologyError) -> Self {
+        MutationError::User(e.to_string())
+    }
+}
+
+impl NetworkTopology {
+    /// Check the integrity of the graph according to all assumptions that we
+    /// make about the graph. This should never fail, and if it does, it's a
+    /// bug.
+    /// - There are no self edges.
+    /// - There are no duplicate edges.
+    /// - There are no edges that reference nodes that are not in the graph.
+    /// - There are no edges that reference conductors that are not equal to
+    ///   their origin node.
+    pub fn integrity_check(&self) -> Result<(), NetworkTopologyError> {
+        for edge in self.edge_references() {
+            // Check that there are no self edges.
+            if edge.source() == edge.target() {
+                return Err(NetworkTopologyError::IntegritySelfEdge);
+            }
+
+            // Check that there are no duplicate edges.
+            let mut edge_set = HashSet::new();
```

Comment:
This must be outside of the loop to be retained across iterations.

Reply:
nice catch!